python/mlc_chat/compiler/quantization/ft_quantization.py

```diff
@@ -0,0 +1,417 @@
+        self,
+        model: nn.Module,
+        quant_map: QuantizeMapping,
+        name_prefix: str,
+    ) -> nn.Module:
+        """
+        Quantize model with group quantization
+
+        Parameters
+        ----------
+        model : nn.Module
+            The non-quantized nn.Module.
+
+        quant_map : QuantizeMapping
+            The quantize mapping with name mapping and func mapping.
+
+        name_prefix : str
+            The name prefix for visited weight.
+
+        Returns
+        -------
+        ret : nn.Module
+            The quantized nn.Module.
+        """
+
+        class _Mutator(nn.Mutator):
+            def __init__(self, config: FTQuantize, quant_map: QuantizeMapping) -> None:
+                super().__init__()
+                self.config = config
+                self.quant_map = quant_map
+
+            def visit_module(self, name: str, node: nn.Module) -> Any:
+                """
+                The visiting method for group quantization of nn.Module nodes.
+
+                Parameters
+                ----------
+                name : str
+                    The name of the current node.
+
+                node : nn.Module
+                    The current node of nn.Module to mutate.
+
+                Returns
+                ------
+                ret_node: Any
+                    The new node to replace current node.
+                """
+                if isinstance(node, nn.Linear):
+                    weight_name = f"{name}.weight"
+                    self.quant_map.param_map[weight_name] = [f"{name}.q_weight", f"{name}.q_scale"]
+                    self.quant_map.map_func[weight_name] = self.config.quantize_weight
+                    return FTQuantizeLinear.from_linear(node, self.config)
+                if isinstance(node, nn.Embedding):
+                    weight_name = f"{name}.weight"
+                    self.quant_map.param_map[weight_name] = [f"{name}.q_weight", f"{name}.q_scale"]
+                    group_quantize = self.config.fallback_group_quantize()
+                    self.quant_map.map_func[weight_name] = group_quantize.quantize_weight
+                    return GroupQuantizeEmbedding.from_embedding(node, group_quantize)
+                return self.visit(name, node)
+
+        model.to(dtype=self.model_dtype)
+        mutator = _Mutator(self, quant_map)
+        model = mutator.visit(name_prefix, model)
+        return model
+
+    def _dequantize(
+        self,
+        weight: te.Tensor,
+        scale: te.Tensor,
+        out_shape: Optional[List[tir.PrimExpr]] = None,
+    ):
+        # cur_group_size = weight.shape[0] if self.group_size == -1 else self.group_size
+        shape = (weight.shape[0], weight.shape[1] * self.num_elem_per_storage)
+        w = te.compute(
+            shape=shape if out_shape is None else out_shape,
+            fcompute=lambda i, j: tir.const(0, self.model_dtype),
+            name="dequantize",
+        )
+        return w
```

Comment:
TODO: Use `nn.SourceModule` instead of dummy decode placeholder for BYOC patterns matching.